File: src/main/java/top/chenray/antilitematica/guard/CommandGuard.java
package top.chenray.antilitematica.guard;

import java.util.Locale;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

import org.bukkit.Bukkit;
import org.bukkit.GameMode;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerCommandPreprocessEvent;
import org.bukkit.event.player.PlayerQuitEvent;

import top.chenray.antilitematica.AntiLitematicaPlugin;
import top.chenray.antilitematica.config.Settings;
import top.chenray.antilitematica.util.Msg;
import top.chenray.antilitematica.util.ViolationWindow;

/**
 * Detects Litematica "quick paste" abuse via rapid command execution
 * (e.g. /setblock spam used by Litematica's executeOperation).
 */
public final class CommandGuard implements Listener {
   private final AntiLitematicaPlugin plugin;
   private final Settings settings;
   private final Map<UUID, Long> lastCommandMs = new ConcurrentHashMap<>();
   private final Map<UUID, Integer> commandBurst = new ConcurrentHashMap<>();
   private final Map<UUID, ViolationWindow> violations = new ConcurrentHashMap<>();

   public CommandGuard(AntiLitematicaPlugin plugin, Settings settings) {
      this.plugin = plugin;
      this.settings = settings;
   }

   public void start() {
      if (this.settings.commandGuard().enabled()) {
         Bukkit.getPluginManager().registerEvents(this, this.plugin);
      }
   }

   public void shutdown() {
      HandlerList.unregisterAll(this);
      this.lastCommandMs.clear();
      this.commandBurst.clear();
      this.violations.clear();
   }

   @EventHandler(
      priority = EventPriority.HIGHEST,
      ignoreCancelled = true
   )
   public void onCommand(PlayerCommandPreprocessEvent event) {
      Player p = event.getPlayer();
      if (shouldBypass(p)) {
         return;
      }
      // Check world whitelist
      if (this.settings.worldWhitelist() != null
            && this.settings.worldWhitelist().isWorldExempt(p.getWorld().getName())) {
         return;
      }

      Settings.CommandGuard cg = this.settings.commandGuard();
      int effectiveMaxPerWindow = plugin.getDynamicThresholdManager().adjustInt(cg.maxPerWindow());
      String cmd = event.getMessage();
      String cmdLower = cmd.toLowerCase(Locale.ROOT);

      // Check allowed commands whitelist (takes priority)
      for (String allowed : cg.allowedCommands()) {
         if (cmdLower.startsWith(allowed.toLowerCase(Locale.ROOT))) {
            return; // Skip all checks for allowed commands
         }
      }

      // Check blocked commands (Litematica quick-paste often uses /setblock)
      boolean blocked = false;
      for (String blockedCmd : cg.blockedCommands()) {
         if (cmdLower.startsWith(blockedCmd.toLowerCase(Locale.ROOT))) {
            blocked = true;
            break;
         }
      }

      // Burst detection: rapid commands in short window
      UUID id = p.getUniqueId();
      long now = System.currentTimeMillis();
      Long last = this.lastCommandMs.put(id, now);
      if (last != null && now - last < cg.windowMs()) {
         int burst = this.commandBurst.merge(id, 1, Integer::sum);
         if (burst >= effectiveMaxPerWindow) {
            this.deny(event, p, "command_burst", burst);
            return;
         }
      } else {
         this.commandBurst.put(id, 1);
      }

      if (blocked) {
         this.deny(event, p, "blocked_command", 1);
      }
   }

   @EventHandler
   public void onQuit(PlayerQuitEvent event) {
      UUID id = event.getPlayer().getUniqueId();
      this.lastCommandMs.remove(id);
      this.commandBurst.remove(id);
      this.violations.remove(id);
   }

   private void deny(PlayerCommandPreprocessEvent event, Player p, String type, int burstCount) {
      event.setCancelled(true);
      ViolationWindow vw = this.violations.computeIfAbsent(p.getUniqueId(),
            ignored -> new ViolationWindow(this.settings.commandGuard().violations().windowMs()));
      int n = vw.addViolation();

      if (this.settings.integration().enabled()) {
         Settings.Integration integ = this.settings.integration();
         this.plugin.getIntegrationManager().flag(p, integ.checkPrefix() + ":command_" + type,
               integ.violationLevel(), "command guard: " + type + " burst=" + burstCount);
      }

      if (n == 1 || n % 2 == 0) {
         String prefix = Msg.prefix(this.settings);
         p.sendMessage(Msg.color(prefix + "&cCommand blocked: suspected Litematica quick-paste abuse."));
      }

      if (n >= plugin.getDynamicThresholdManager().adjustInt(this.settings.commandGuard().violations().kickAt())) {
         if (p.isOnline()) {
            this.plugin.getLogger().info("Kicking " + p.getName() + " due to repeated blocked commands (" + type + "), violations=" + n);
            Bukkit.getScheduler().runTask(this.plugin, () -> {
               if (p.isOnline() && !shouldBypass(p)) {
                  String prefix = Msg.prefix(this.settings);
                  p.kickPlayer(Msg.color(prefix + this.settings.messages().kick()));
               }
            });
         }
      } else {
         this.notifyStaff(p, type, n);
      }
   }

   private void notifyStaff(Player p, String type, int violations) {
      String prefix = Msg.prefix(this.settings);
      String msg = Msg.color(prefix + "&e" + p.getName() + " &7blocked command (&f" + type + "&7), vio=&f" + violations);
      for (Player online : Bukkit.getOnlinePlayers()) {
         if (online.hasPermission("antilitematica.notify")) {
            online.sendMessage(msg);
         }
      }
   }

   private static boolean shouldBypass(Player p) {
      return p.hasPermission("antilitematica.bypass");
   }
}

File: src/main/kotlin/icu/epochcraft/antilitematica/guard/CommandGuard.kt
package icu.epochcraft.antilitematica.guard

import icu.epochcraft.antilitematica.AntiLitematica
import icu.epochcraft.antilitematica.event.DetectionType
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerCommandPreprocessEvent
import org.bukkit.event.player.PlayerQuitEvent
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * 命令防护：拦截 Litematica quick-paste 命令滥用（如 /setblock 连发）。
 *
 * 逻辑（移植自旧版 CommandGuard，修复 B6）：
 *   - allowed 白名单优先（精确匹配或前缀+空格，不误伤 /setblockinfo 这类）
 *   - 命中 blocked 列表的命令：窗口内 burst 限流，超限判定
 *   - 惩罚统一走检测总线（不再自实现踢人）
 *
 * 默认 blocked 列表为空——由管理员按服配置，避免误伤（旧版默认禁 /setblock 全服）。
 *
 * @author 阿清
 */
class CommandGuard(private val plugin: AntiLitematica) : Listener {

    // ---- 缓存配置 ----
    private var enabled = false
    private var allowedCommands: List<String> = emptyList()
    private var blockedCommands: List<String> = emptyList()
    private var maxPerWindow = 8
    private var windowMs = 2000L

    /** world -> player -> 上次命令时间 */
    private val lastCommandMs = ConcurrentHashMap<String, MutableMap<UUID, Long>>()
    /** world -> player -> 窗口内 burst 计数 */
    private val burstCount = ConcurrentHashMap<String, MutableMap<UUID, Int>>()

    /** reload 配置缓存 */
    fun reload() {
        val cg = plugin.configHolder.commandGuard
        enabled = cg.enabled
        allowedCommands = cg.allowedCommands
        blockedCommands = cg.blockedCommands
        maxPerWindow = cg.maxPerWindow.coerceAtLeast(1)
        windowMs = cg.windowMs.coerceAtLeast(500)
        if (!enabled) {
            lastCommandMs.clear()
            burstCount.clear()
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onCommand(event: PlayerCommandPreprocessEvent) {
        if (!enabled) return
        val p = event.player
        if (p.hasPermission("antilitematica.bypass")) return

        val cmdLower = event.message.lowercase()
        val world = p.world.name.lowercase()

        // 1. allowed 白名单优先（精确匹配或 "命令 " 前缀，避免误伤子命令）
        if (allowedCommands.any { a -> cmdLower == a || cmdLower.startsWith("$a ") }) return

        // 2. 是否命中 blocked 列表
        val blocked = blockedCommands.any { b -> cmdLower == b || cmdLower.startsWith("$b ") }
        if (!blocked) return

        // 3. 窗口内 burst 限流
        val now = System.currentTimeMillis()
        val cmdMap = lastCommandMs.computeIfAbsent(world) { ConcurrentHashMap() }
        val burstMap = burstCount.computeIfAbsent(world) { ConcurrentHashMap() }

        val last = cmdMap.put(p.uniqueId, now)
        val burst: Int = if (last != null && now - last < windowMs) {
            burstMap.merge(p.uniqueId, 1, Int::plus) ?: 1
        } else {
            burstMap[p.uniqueId] = 1
            1
        }

        // 连续快速执行 = quick-paste 特征
        if (burst >= maxPerWindow) {
            deny(event, p, "command_burst")
            return
        }
        deny(event, p, "blocked_command")
    }

    /** 判定命中：取消命令并发射到检测总线 */
    private fun deny(event: PlayerCommandPreprocessEvent, p: Player, type: String) {
        event.isCancelled = true
        plugin.detectionBus.emit(
            p,
            "command:$type",
            "命令滥用: $type",
            DetectionType.COMMAND,
        )
    }

    @EventHandler
    fun onQuit(event: PlayerQuitEvent) {
        val id = event.player.uniqueId
        lastCommandMs.values.forEach { it.remove(id) }
        burstCount.values.forEach { it.remove(id) }
    }
}
